Assets/MRTK/SDK/Features/UX/Scripts/Pointers/SpherePointer.cs

```diff
@@ -202,27 +218,49 @@ public override void OnPreSceneQuery()
                 {
                     Vector3 endPoint = Vector3.forward * SphereCastRadius;
                     Rays[0].UpdateRayStep(ref pointerPosition, ref endPoint);
-                    PrioritizedLayerMasksOverride = PrioritizedLayerMasksOverride ?? GrabLayerMasks;
 
                     for (int i = 0; i < PrioritizedLayerMasksOverride.Length; i++)
                     {
-                        if (queryBufferNearObjectRadius.TryUpdateQueryBufferForLayerMask(PrioritizedLayerMasksOverride[i], pointerPosition - pointerAxis * PullbackDistance, pointerAxis, triggerInteraction, ignoreCollidersNotInFOV))
+                        // First update queryBufferNearObjectRadius to see if there is a grabbable in the near interaction range
+                        queryBufferNearObjectRadius.TryUpdateQueryBufferForLayerMask(PrioritizedLayerMasksOverride[i], pointerPosition - pointerAxis * PullbackDistance, triggerInteraction);
+                        if (queryBufferNearObjectRadius.HasValidGrabbable(pointerPosition - pointerAxis * PullbackDistance, pointerAxis, ignoreCollidersNotInFOV))
                         {
                             break;
                         }
                     }
 
+                    hitObject = null;
+                    hitPoint = Vector3.zero;
+                    hitDistance = Mathf.Infinity;
                     for (int i = 0; i < PrioritizedLayerMasksOverride.Length; i++)
                     {
-                        if (queryBufferInteractionRadius.TryUpdateQueryBufferForLayerMask(PrioritizedLayerMasksOverride[i], pointerPosition, pointerAxis, triggerInteraction, ignoreCollidersNotInFOV))
+                        // Then update queryBufferInteractionRadius to see if there is a grabbable that can be interacted with
+                        queryBufferInteractionRadius.TryUpdateQueryBufferForLayerMask(PrioritizedLayerMasksOverride[i], pointerPosition, triggerInteraction);
+                        if (queryBufferInteractionRadius.HasValidGrabbable(pointerPosition, pointerAxis, ignoreCollidersNotInFOV))
                         {
-                            break;
+                            hitObject = queryBufferInteractionRadius.GetClosestValidGrabbable(pointerPosition, pointerAxis, IgnoreCollidersNotInFOV, out hitPoint);
```

Comment:
This is maybe an optimization we could make longer term, but I wonder if we even need to make the second `TryUpdateQueryBufferForLayerMask` with the smaller radius call?

Since `GetClosestValidGrabbable` calculates the distance to each grabbable, would it be less expensive to compare that distance against the smaller radius to filter out anything too far?

Reply:
I _think_ the second OverlapSphere call is overall less expensive than calculating the extra distances, since each of those distance calculations will end up collider.ClosestPoint(). The size of the space covered by the NearObjectCall grows at r^3, where r is the radius we specified, so I think this makes the set of potential objects captured by the overlap sphere call also scale in a less than desirable way. 

Additionally, since queryBufferNearObjectRadius only checks for the existence of an object that fits within the query angle, it only hits O(n) performance in the worst case, whereas if it also had to filter down and find the object that was the smallest distance away, it would always be O(n). Combined with the number of viable objects increasing at a possibly greater than linear rate with the initial sphere cast radius, I think it's probably more performant to do two separate sphere cast calls, with the InteractionRadius call being the smaller for the two). 

It also makes sense that we do the InteractionRadius call even when the NearObject call doesn't succeed, since there are edge cases where an object is grabbable (i.e. our hand is slightly in front of it's grabbable point), but not in the "near interaction" detection range (which we might define as being within a certain distance and angle of the hand)